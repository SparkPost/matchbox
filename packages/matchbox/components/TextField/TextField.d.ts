export default TextField;
declare const TextField: any;
