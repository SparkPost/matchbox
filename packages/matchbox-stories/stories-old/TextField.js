import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { action } from '@storybook/addon-actions';
import { StoryContainer } from './helpers';

import { TextField, Button, Select, Tooltip } from '../src';
import { Autorenew } from '@sparkpost/matchbox-icons';

export default storiesOf('TextField', module)
  .addDecorator((getStory) => (
    <StoryContainer bg='white'>{ getStory() }</StoryContainer>
  ))
  .add('Default', withInfo()(() => (
    <TextField
      id='id'
      label='Name'
      placeholder='Leslie Knope'
    />
  )))

  .add('with an error', withInfo()(() => (
    <TextField
      id='id'
      label='Name'
      error='You forgot your name!'
    />
  )))

  .add('with an inline error', withInfo()(() => (
    <TextField
      id='id'
      label='Name'
      inlineErrors={true}
      error='You forgot my name!'
    />
  )))

  .add('multiline', withInfo()(() => (
    <TextField
      id='id'
      label='Your Message'
      multiline
    />
  )))

  .add('disabled', withInfo()(() => (
    <TextField
      id='id'
      label='Template ID'
      value='template-12'
      disabled
    />
  )))

  .add('required', withInfo()(() => (
    <TextField
      id='id'
      label='Template ID'
      value='template-12'
      required
    />
  )))

  .add('with help text', withInfo()(() => (
      <TextField
        id='id'
        label='Template ID'
        helpText='A unique ID for your template.'
      />
  )))

  .add('with connected components', withInfo()(() => (
      <TextField
        id='id'
        label='Date Range'
        value='July 21, 2017 - July 28, 2017'
        connectLeft={<Tooltip content='Hey'><Button>Injection Time</Button></Tooltip>}
        connectRight={<Select options={
          ['Last 24 Hours', 'Last Week']
        }/>}
      />
  )))

  .add('with prefix and suffix', withInfo()(() => (
      <TextField
        id='id'
        prefix='$'
        suffix={<Autorenew />}
      />
  )));
