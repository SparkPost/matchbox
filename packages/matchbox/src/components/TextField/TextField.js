import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Label } from '../Label';
import { Error } from '../Error';
import { Connect } from '../Connect';
import classnames from 'classnames';
import styles from './TextField.module.scss';

class TextField extends Component {
  static displayName = 'TextField';

  static propTypes = {
    /**
     * 'left' | 'center' | 'right'
     */
    align: PropTypes.oneOf(['left', 'center', 'right']),
    id: PropTypes.string,
    multiline: PropTypes.bool,
    /**
     * 'text' | 'email' | 'number' | 'password' | 'search' | 'tel' | 'url' | 'date'
     */
    type: PropTypes.oneOf(['text', 'email', 'number', 'password', 'search', 'tel', 'url', 'date']),
    placeholder: PropTypes.string,
    autoFocus: PropTypes.bool,
    disabled: PropTypes.bool,
    readOnly: PropTypes.bool,
    required: PropTypes.bool,
    /**
     * 'none' | 'both' | 'horizontal' | 'vertical' | 'block' | 'inline'
     */
    resize: PropTypes.oneOf(['none', 'both', 'horizontal', 'vertical', 'block', 'inline']),
    label: PropTypes.string,
    labelHidden: PropTypes.bool,
    helpText: PropTypes.node,
    connectLeft: PropTypes.node,
    connectRight: PropTypes.node,
    prefix: PropTypes.node,
    suffix: PropTypes.node,
    error: PropTypes.string,
    errorInLabel: PropTypes.bool,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func
  };

  static defaultProps = {
    resize: 'both',
    type: 'text'
  };

  state = {
    paddingLeft: null,
    paddingRight: null
  };

  componentDidMount() {
    this.calculatePadding();
  }

  componentDidUpdate({ prefix, suffix }) {
    if (prefix !== this.props.prefix || suffix !== this.props.suffix) {
      this.calculatePadding();
    }
  }

  calculatePadding() {
    this.setState({
      paddingLeft: this.prefixNode && this.prefixNode.getBoundingClientRect().width + 20,
      paddingRight: this.suffixNode && this.suffixNode.getBoundingClientRect().width + 20
    });
  }

  render() {
    const {
      align,
      id,
      name,
      multiline,
      type,
      placeholder,
      autoFocus,
      disabled,
      readOnly,
      required,
      resize,
      label,
      labelHidden,
      helpText,
      error,
      errorInLabel,
      onChange,
      onFocus,
      onBlur,
      connectLeft,
      connectRight,
      prefix,
      prefixClassname,
      suffix,
      suffixClassname,
      style,
      ...rest
    } = this.props;

    const { paddingLeft, paddingRight } = this.state;

    const setInputClasses = classnames(
      styles.Input,
      error && styles.Error
    );

    const prefixClasses = classnames(styles.Prefix, prefixClassname);
    const suffixClasses = classnames(styles.Suffix, suffixClassname);

    const requiredIndicator = required
      ? ' *'
      : '';

    const labelMarkup = (
      <Label id={id} label={`${label}${requiredIndicator}`}>
        {error && errorInLabel && <Error className={styles.InlineError} wrapper='span' error={error} />}
      </Label>
    );

    const helpMarkup = helpText
      ? <div className={styles.HelpText}>{helpText}</div>
      : null;

    const prefixMarkup = prefix
      ? <span className={prefixClasses} ref={(node) => this.prefixNode = node}>{prefix}</span>
      : null;

    const suffixMarkup = suffix
      ? <span className={suffixClasses} ref={(node) => this.suffixNode = node}>{suffix}</span>
      : null;

    const input = React.createElement(multiline ? 'textarea' : 'input', {
      ...rest,
      name,
      id,
      type,
      disabled,
      readOnly,
      autoFocus,
      placeholder,
      onFocus,
      onBlur,
      onChange,
      className: setInputClasses,
      style: {
        textAlign: align,
        paddingLeft,
        paddingRight,
        resize,
        ...style
      }
    });

    const inputMarkup = prefix || suffix
      ? <div className={styles.InputWrapper}>
        {prefixMarkup}
        {input}
        {suffixMarkup}
      </div>
      : input;

    return (
      <fieldset className={styles.TextField}>
        {label && !labelHidden && labelMarkup}
        <Connect left={connectLeft} right={connectRight}>
          {inputMarkup}
        </Connect>
        {error && !errorInLabel && <Error error={error} />}
        {helpMarkup}
      </fieldset>
    );
  }
}

export default TextField;
