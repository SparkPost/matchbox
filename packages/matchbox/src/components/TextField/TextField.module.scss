@import '../../styles/config';

.TextField {
  position: relative;
  border: none;
  padding: 0;
  margin: 0 0 spacing() 0;

  &:last-child {
    margin: 0;
  }
}

.InlineError {
  padding-left: spacing(small);
}

.Input {
  position: relative;
  display: block;
  box-sizing: border-box;
  width: 100%;

  padding: rem(5) rem(12);

  outline: none;
  border: 1px solid color(gray, 7);
  border-radius: border-radius();

  color: color(gray, 1);
  font-weight: 400;
  font-size: font-size(400);
  line-height: rem(24);

  transition: border 0.15s;

  &.Error {
    border: 1px solid color(red);
    background: rgba(color(red, dark), 0.06);
  }

  &:focus {
    border: 1px solid color(blue);
    box-shadow: 0 0 0 1px color(blue);
  }

  &:disabled {
    background: color(gray, 9);
    color: color(gray, 3);
    cursor: not-allowed;
  }
}

.HelpText {
  padding-top: rem(6);
  font-size: font-size(300);
  line-height: line-height(300);
  color: color(gray, 4);
  word-break: normal;
  word-wrap: normal;
}

.Prefix, .Suffix {
  display: inline-block;
  position: absolute;
  top: 0;
  z-index: 2;
  pointer-events: none;
  opacity: 0.5;
}

.Prefix {
  left: rem(12);
  padding: rem(6) 0;
}

.Suffix {
  right: rem(12);
  padding: rem(6) 0;
}

.InputWrapper {
  position: relative;
}
