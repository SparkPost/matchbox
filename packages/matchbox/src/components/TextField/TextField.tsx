import React, { useRef } from 'react';
import { tokens } from '@sparkpost/design-tokens';
import { compose, margin, MarginProps, maxWidth, MaxWidthProps } from 'styled-system';
import { omit } from '@styled-system/props';
import styled from 'styled-components';
import { Label } from '../Label';
import { ErrorLabel } from '../Error';
import { Connect } from '../Connect';
import { Box, BoxProps } from '../Box';
import { OptionalLabel } from '../OptionalLabel';
import { HelpText } from '../HelpText';
import { KeyboardArrowDown, KeyboardArrowUp } from '@sparkpost/matchbox-icons';
import { roundToBaseline } from '../../helpers/geometry';
import useInputDescribedBy from '../../hooks/useInputDescribedBy';
import useResizeObserver from '../../hooks/useResizeObserver';
import { pick } from '../../helpers/props';
import { focusOutline, buttonReset } from '../../styles/helpers';
import type { ComponentPropsWithout } from '../../helpers/types';
import { combineRefs } from '../../helpers/ref';

const system = compose(margin, maxWidth);
const StyledWrapper = styled('div')`
  ${system}
`;

const StyledInputWrapper = styled(Box)`
  ${focusOutline({ modifier: '&:focus-within', offset: '2px' })}
`;

const StyledInput = styled(Box)<BoxProps>`
  outline: none;
  &[type='number'],
  &[type='number']::-webkit-inner-spin-button,
  &[type='number']::-webkit-outer-spin-button {
    appearance: none;
    margin: 0;
  }
`;
const StyledButton = styled.button<
  React.ComponentPropsWithRef<'button'> & Omit<React.ComponentProps<typeof Box>, 'as'>
>`
  ${buttonReset}
  cursor: pointer;
`;

const StyledNumberControls = styled(Box)<BoxProps & { $disabled?: boolean }>`
  padding: 0 6px;
  position: absolute;
  right: 0;
  top: 0;
  height: 100%;
  border-left: ${(props) => props.theme.borders['400']};
  display: flex;
  flex-direction: column;
`;

const StyledArrowUp = styled(KeyboardArrowUp)`
  color: ${(props) => props.theme.colors.blue['700']};
  display: block;
`;

const StyledArrowDown = styled(KeyboardArrowDown)`
  color: ${(props) => props.theme.colors.blue['700']};
  display: block;
`;

type FieldBoxProps = BoxProps &
  React.ComponentPropsWithRef<'input'> & {
    hasError?: boolean;
  };

const FieldBox = React.forwardRef<HTMLInputElement, FieldBoxProps>(function FieldBox(
  props,
  userRef,
) {
  const { hasError, disabled, height, type, lineHeight, required, py, ...rest } = props;

  const inputRef = useRef<HTMLInputElement>(null);

  const increment = () => {
    inputRef.current.stepUp();
  };
  const decrement = () => {
    inputRef.current.stepDown();
  };

  return (
    <StyledInputWrapper>
      <StyledInput
        aria-invalid={!!hasError}
        as="input"
        px="400"
        {...rest}
        type={type}
        disabled={disabled}
        width="100%"
        border={hasError ? `1px solid ${tokens.color_red_700}` : '400'}
        borderRadius="100"
        bg={disabled ? 'gray.200' : 'white'}
        color="gray.900"
        height={height}
        lineHeight={lineHeight}
        py={py}
        required={required}
        ref={combineRefs(inputRef, userRef)}
      />
      {type == 'number' && !disabled && (
        <StyledNumberControls $disabled={disabled}>
          <StyledButton
            onClick={increment}
            as="button"
            type="button"
            aria-hidden={true}
            disabled={disabled}
            tabIndex={-1}
          >
            <StyledArrowUp size={20} />
          </StyledButton>
          <StyledButton
            onClick={decrement}
            as="button"
            type="button"
            aria-hidden={true}
            disabled={disabled}
            tabIndex={-1}
          >
            <StyledArrowDown size={20} />
          </StyledButton>
        </StyledNumberControls>
      )}
    </StyledInputWrapper>
  );
});

type PrefixOrSuffixProps = BoxProps &
  React.ComponentPropsWithRef<'span'> & {
    content?: React.ReactNode;
  };

const PrefixOrSuffix = React.forwardRef<HTMLSpanElement, PrefixOrSuffixProps>(
  function PrefixOrSuffix(props, ref) {
    const { content, className, ...rest } = props;
    if (!content) {
      return null;
    }

    return (
      <Box
        as="span"
        position="absolute"
        zIndex="default"
        top="0"
        lineHeight="2.5rem"
        height="2.5rem"
        className={className}
        ref={ref}
        {...rest}
      >
        {content}
      </Box>
    );
  },
);

type SharedTextFieldProps = {
  id: string; // Required
  align?: 'left' | 'center' | 'right';
  connectLeft?: React.ReactNode;
  connectRight?: React.ReactNode;
  error?: React.ReactNode;
  errorInLabel?: boolean;
  helpText?: React.ReactNode;
  label?: React.ReactNode;
  labelHidden?: boolean;
  prefix?: React.ReactNode;
  prefixClassname?: string;
  suffix?: React.ReactNode;
  suffixClassname?: string;
  resize?: 'none' | 'both' | 'horizontal' | 'vertical' | 'block' | 'inline';
  type?: string;
  optional?: boolean;
} & MarginProps &
  MaxWidthProps;

// Omitting 'prefix' because it is a native html attribute
type MultiLineProps = ComponentPropsWithout<'textarea', 'prefix'> &
  SharedTextFieldProps & {
    multiline: true;
  };

// Omitting 'prefix' because it is a native html attribute
type InputProps = ComponentPropsWithout<'input', 'prefix'> &
  SharedTextFieldProps & {
    multiline?: false;
  };

export type TextFieldProps = MultiLineProps | InputProps;

const TextField = React.forwardRef<HTMLInputElement | HTMLTextAreaElement, TextFieldProps>(
  function TextField(props, userRef) {
    const {
      align,
      autoFocus,
      connectLeft,
      connectRight,
      disabled,
      error,
      errorInLabel,
      helpText,
      id,
      label,
      labelHidden,
      multiline,
      name,
      onChange,
      onFocus,
      onBlur,
      placeholder,
      prefix,
      prefixClassname,
      readOnly,
      required,
      resize = 'both',
      type = 'text',
      suffix,
      suffixClassname,
      style,
      optional,
      ...rest
    } = props;

    const componentProps = omit(rest);
    const systemProps = pick(rest, system.propNames);
    const [prefixRef, prefixEntry] = useResizeObserver();
    const [suffixRef, suffixEntry] = useResizeObserver();
    const [prefixPadding, setPrefixPadding] = React.useState(0);
    const [suffixPadding, setSuffixPadding] = React.useState(0);
    const { describedBy, errorId, helpTextId } = useInputDescribedBy({
      id,
      hasHelpText: !!helpText,
      hasError: !!error,
    });

    React.useLayoutEffect(() => {
      if (prefix && prefixEntry.contentRect.width) {
        setPrefixPadding(roundToBaseline(prefixEntry.contentRect.width));
      }
    }, [prefix, prefixEntry]);

    React.useLayoutEffect(() => {
      if (suffix && suffixEntry.contentRect.width) {
        setSuffixPadding(roundToBaseline(suffixEntry.contentRect.width));
      }
    }, [suffix, suffixEntry]);

    const inputProps = {
      ...componentProps,
      as: multiline ? 'textarea' : 'input',
      height: multiline ? 'auto' : '2.5rem',
      lineHeight: multiline ? '300' : '2.5rem',
      py: multiline ? '300' : '0',
      hasError: !!error,
      name,
      id,
      type,
      disabled,
      readOnly,
      required,
      autoFocus,
      placeholder,
      onFocus,
      onBlur,
      onChange,
      pl: `calc(${prefixPadding}px + ${tokens.spacing_400})`,
      pr: `calc(${suffixPadding}px + ${tokens.spacing_400})`,
      textAlign: align,
      ...describedBy,
      style: {
        cursor: disabled ? 'not-allowed' : 'auto',
        ...(multiline ? { resize } : {}),
        ...style,
      },
    };

    return (
      <StyledWrapper {...systemProps}>
        {label && (
          <Label id={id} labelHidden={labelHidden}>
            <Box as="span" pr="200">
              {label}
            </Box>
            {error && errorInLabel && (
              <ErrorLabel id={errorId} wrapper="span">
                {error}
              </ErrorLabel>
            )}
            {optional && <OptionalLabel float />}
          </Label>
        )}
        <Connect left={connectLeft} right={connectRight}>
          <Box position="relative">
            <PrefixOrSuffix
              content={prefix}
              className={prefixClassname}
              ref={prefixRef}
              left="300"
            />
            <FieldBox ref={userRef} {...inputProps} />
            <PrefixOrSuffix
              content={suffix}
              className={suffixClassname}
              ref={suffixRef}
              right="300"
            />
          </Box>
        </Connect>
        {helpText && <HelpText id={helpTextId}>{helpText}</HelpText>}
        {error && !errorInLabel && <ErrorLabel id={errorId}>{error}</ErrorLabel>}
      </StyledWrapper>
    );
  },
);

TextField.displayName = 'TextField';
export default TextField;
