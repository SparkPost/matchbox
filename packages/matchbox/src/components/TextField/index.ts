export { default as TextField } from './TextField';
export * from './TextField';
