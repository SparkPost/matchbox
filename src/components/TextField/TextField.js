import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Label } from '../Label';
import { Error } from '../Error';
import { Connect } from '../Connect';
import classnames from 'classnames';
import styles from './TextField.module.scss';

const InlineErrorWrapper = ({ children }) => <span className={styles.InlineError}>{children}</span>;

class TextField extends Component {
  static displayName = 'TextField';

  static propTypes = {
    id: PropTypes.string,
    multiline: PropTypes.bool,
    /**
     * 'text' | 'email' | 'number' | 'password' | 'search' | 'tel' | 'url' | 'date'
     */
    type: PropTypes.oneOf(['text', 'email', 'number', 'password', 'search', 'tel', 'url', 'date']),
    placeholder: PropTypes.string,
    autoFocus: PropTypes.bool,
    disabled: PropTypes.bool,
    readOnly: PropTypes.bool,
    required: PropTypes.bool,
    /**
     * 'none' | 'both' | 'horizontal' | 'vertical' | 'block' | 'inline'
     */
    resize: PropTypes.oneOf(['none', 'both', 'horizontal', 'vertical', 'block', 'inline']),
    label: PropTypes.string,
    labelHidden: PropTypes.bool,
    helpText: PropTypes.node,
    connectLeft: PropTypes.node,
    connectRight: PropTypes.node,
    prefix: PropTypes.node,
    suffix: PropTypes.node,
    error: PropTypes.string,
    /**
     * Inlines the error message next to the field label. Label prop required.
     */
    inlineErrors: PropTypes.bool,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func
  };

  static defaultProps = {
    required: false,
    resize: 'both',
    type: 'text'
  };

  state = {
    paddingLeft: null,
    paddingRight: null
  };

  componentDidMount() {
    this.calculatePadding();
  }

  componentDidUpdate({ prefix, suffix }) {
    if (prefix !== this.props.prefix || suffix !== this.props.suffix) {
      this.calculatePadding();
    }
  }

  calculatePadding() {
    this.setState({
      paddingLeft: this.prefixNode && this.prefixNode.getBoundingClientRect().width + 20,
      paddingRight: this.suffixNode && this.suffixNode.getBoundingClientRect().width + 20
    });
  }

  render() {
    const {
      id,
      name,
      multiline,
      type,
      placeholder,
      autoFocus,
      disabled,
      readOnly,
      required,
      resize,
      label,
      labelHidden,
      helpText,
      inlineErrors,
      error,
      onChange,
      onFocus,
      onBlur,
      connectLeft,
      connectRight,
      prefix,
      suffix,
      ...rest
    } = this.props;

    const { paddingLeft, paddingRight } = this.state;

    const setClasses = classnames(
      styles.TextField,
      error && styles.error
    );

    const requiredIndicator = required
      ? ' *'
      : '';

    let labelMarkup;
    const errorWrapper = inlineErrors ? InlineErrorWrapper : 'div';
    const renderLabel = label && (!labelHidden || inlineErrors);
    if (renderLabel) {
      const renderErrorInLabel = error && inlineErrors;
      const errorMsg = renderErrorInLabel ? <Error wrapper={errorWrapper} error={error} /> : null;
      labelMarkup = <Label
        id={id}
        label={`${label}${requiredIndicator}`}>
        {errorMsg}
      </Label>;
    }

    const errorMarkup = error && !inlineErrors
      ? <Error error={error} />
      : null;

    const helpMarkup = helpText
      ? <div className={styles.HelpText}>{ helpText }</div>
      : null;

    const prefixMarkup = prefix
      ? <span className={styles.Prefix} ref={(node) => this.prefixNode = node}>{ prefix }</span>
      : null;

    const suffixMarkup = suffix
      ? <span className={styles.Suffix} ref={(node) => this.suffixNode = node}>{ suffix }</span>
      : null;

    const input = React.createElement(multiline ? 'textarea' : 'input', {
      ...rest,
      name,
      id,
      type,
      disabled,
      readOnly,
      autoFocus,
      placeholder,
      onFocus,
      onBlur,
      onChange,
      className: styles.Input,
      style: { paddingLeft, paddingRight, resize }
      // 'aria-describedby':
      // 'aria-labelledby':
      // 'aria-invalid':
    });

    const inputMarkup = prefix || suffix
      ? <div className={styles.InputWrapper}>
          { prefixMarkup }
          { input }
          { suffixMarkup }
        </div>
      : input;

    return (
      <fieldset className={setClasses}>
        { labelMarkup }
        <Connect left={connectLeft} right={connectRight}>
          { inputMarkup }
        </Connect>
        { errorMarkup }
        { helpMarkup }
      </fieldset>
    );
  }
}

export default TextField;
