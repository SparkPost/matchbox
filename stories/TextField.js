import React from 'react';
import { storiesOf, action, linkTo } from '@storybook/react';
import { StoryContainer } from './helpers';

import { TextField } from '../src';

export default storiesOf('TextField', module)

  .add('Default', () => (
    <StoryContainer bg='white'>
      <TextField
        id='id'
        label='Name'
        placeholder='Leslie Knope'
      />
    </StoryContainer>
  ))

  .add('with an error', () => (
    <StoryContainer bg='white'>
      <TextField
        id='id'
        label='Name'
        error='You forgot your name!'
      />
    </StoryContainer>
  ))

  .add('multiline', () => (
    <StoryContainer bg='white'>
      <TextField
        id='id'
        label='Your Message'
        multiline
      />
    </StoryContainer>
  ));
