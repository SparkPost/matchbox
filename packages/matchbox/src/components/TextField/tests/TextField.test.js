import React from 'react';
import { shallow } from 'enzyme';
import cases from 'jest-in-case';

import TextField from '../TextField';

describe('TextField', () => {
  let wrapper;

  beforeEach(() => {
    const props = {
    };

    wrapper = shallow(<TextField {...props} />);
  });

  const testCases = [
    { name: 'disabled', props: { disabled: true }},
    { name: 'label with required', props: { required: true, label: 'TextField Label' }},
    { name: 'label without required', props: { label: 'TextField Label' }},
    { name: 'hidden label', props: { label: 'TextField Label', labelHidden: true }},
    { name: 'helpText', props: { HelpText: 'TextField help text' }},
    { name: 'error', props: { error: 'An error occurred' }},
    { name: 'errorInLabel', props: { error: 'Error occurred', errorInLabel: true }},
    { name: 'with placehoders', props: { placehodler: 'Enter your email' }},
    { name: 'type email', props: { type: 'email' }},
    { name: 'type password', props: { type: 'password' }},
    { name: 'with multiline', props: { multiline: true }},
    { name: 'with style', props: { style: { textAlign: 'right' }}},
    { name: 'with prefix', props: { prefix: <span>Before Text</span>, prefixClassname: 'prefixclass' }},
    { name: 'with suffix', props: { suffix: <span>After Text</span>, suffixClassname: 'prefixclass' }},
    { name: 'with prefix & suffix', props: { prefix: <span>Before Text</span>, suffix: <span>After Text</span> }},
    { name: 'centered', props: { align: 'center' }},
    { name: 'right', props: { align: 'right' }}
  ];

  cases('renders TextField', (opts) => {
    wrapper.setProps(opts.props);
    expect(wrapper).toMatchSnapshot();
  }, testCases);

  cases('invokes event', (opts) => {
    const fn = jest.fn();
    const newProps = {};
    newProps[opts.name] = fn;
    wrapper.setProps(newProps);

    wrapper.find('input').simulate(opts.event, opts.args); //not testing multiline
    expect(fn).toHaveBeenCalledTimes(1);
  }, [
    { name: 'onChange', event: 'change' },
    { name: 'onBlur', event: 'blur' },
    { name: 'onFocus', event: 'focus' }
  ]);

});
