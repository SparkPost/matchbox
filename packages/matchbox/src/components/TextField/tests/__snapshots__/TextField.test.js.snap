// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField renders TextField centered 1`] = `
<fieldset
  className="TextField"
>
  <Connect>
    <input
      className="Input"
      style={
        Object {
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "resize": "both",
          "textAlign": "center",
        }
      }
      type="text"
    />
  </Connect>
</fieldset>
`;

exports[`TextField renders TextField disabled 1`] = `
<fieldset
  className="TextField"
>
  <Connect>
    <input
      className="Input"
      disabled={true}
      style={
        Object {
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "resize": "both",
          "textAlign": undefined,
        }
      }
      type="text"
    />
  </Connect>
</fieldset>
`;

exports[`TextField renders TextField error 1`] = `
<fieldset
  className="TextField"
>
  <Connect>
    <input
      className="Input Error"
      style={
        Object {
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "resize": "both",
          "textAlign": undefined,
        }
      }
      type="text"
    />
  </Connect>
  <Error
    error="An error occurred"
  />
</fieldset>
`;

exports[`TextField renders TextField errorInLabel 1`] = `
<fieldset
  className="TextField"
>
  <Connect>
    <input
      className="Input Error"
      style={
        Object {
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "resize": "both",
          "textAlign": undefined,
        }
      }
      type="text"
    />
  </Connect>
</fieldset>
`;

exports[`TextField renders TextField helpText 1`] = `
<fieldset
  className="TextField"
>
  <Connect>
    <input
      HelpText="TextField help text"
      className="Input"
      style={
        Object {
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "resize": "both",
          "textAlign": undefined,
        }
      }
      type="text"
    />
  </Connect>
</fieldset>
`;

exports[`TextField renders TextField hidden label 1`] = `
<fieldset
  className="TextField"
>
  <Connect>
    <input
      className="Input"
      style={
        Object {
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "resize": "both",
          "textAlign": undefined,
        }
      }
      type="text"
    />
  </Connect>
</fieldset>
`;

exports[`TextField renders TextField label with required 1`] = `
<fieldset
  className="TextField"
>
  <Label
    label="TextField Label *"
  />
  <Connect>
    <input
      className="Input"
      style={
        Object {
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "resize": "both",
          "textAlign": undefined,
        }
      }
      type="text"
    />
  </Connect>
</fieldset>
`;

exports[`TextField renders TextField label without required 1`] = `
<fieldset
  className="TextField"
>
  <Label
    label="TextField Label"
  />
  <Connect>
    <input
      className="Input"
      style={
        Object {
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "resize": "both",
          "textAlign": undefined,
        }
      }
      type="text"
    />
  </Connect>
</fieldset>
`;

exports[`TextField renders TextField right 1`] = `
<fieldset
  className="TextField"
>
  <Connect>
    <input
      className="Input"
      style={
        Object {
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "resize": "both",
          "textAlign": "right",
        }
      }
      type="text"
    />
  </Connect>
</fieldset>
`;

exports[`TextField renders TextField type email 1`] = `
<fieldset
  className="TextField"
>
  <Connect>
    <input
      className="Input"
      style={
        Object {
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "resize": "both",
          "textAlign": undefined,
        }
      }
      type="email"
    />
  </Connect>
</fieldset>
`;

exports[`TextField renders TextField type password 1`] = `
<fieldset
  className="TextField"
>
  <Connect>
    <input
      className="Input"
      style={
        Object {
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "resize": "both",
          "textAlign": undefined,
        }
      }
      type="password"
    />
  </Connect>
</fieldset>
`;

exports[`TextField renders TextField with multiline 1`] = `
<fieldset
  className="TextField"
>
  <Connect>
    <textarea
      className="Input"
      style={
        Object {
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "resize": "both",
          "textAlign": undefined,
        }
      }
      type="text"
    />
  </Connect>
</fieldset>
`;

exports[`TextField renders TextField with placehoders 1`] = `
<fieldset
  className="TextField"
>
  <Connect>
    <input
      className="Input"
      placehodler="Enter your email"
      style={
        Object {
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "resize": "both",
          "textAlign": undefined,
        }
      }
      type="text"
    />
  </Connect>
</fieldset>
`;

exports[`TextField renders TextField with prefix & suffix 1`] = `
<fieldset
  className="TextField"
>
  <Connect>
    <div
      className="InputWrapper"
    >
      <span
        className="Prefix"
      >
        <span>
          Before Text
        </span>
      </span>
      <input
        className="Input"
        style={
          Object {
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "resize": "both",
            "textAlign": undefined,
          }
        }
        type="text"
      />
      <span
        className="Suffix"
      >
        <span>
          After Text
        </span>
      </span>
    </div>
  </Connect>
</fieldset>
`;

exports[`TextField renders TextField with prefix 1`] = `
<fieldset
  className="TextField"
>
  <Connect>
    <div
      className="InputWrapper"
    >
      <span
        className="Prefix prefixclass"
      >
        <span>
          Before Text
        </span>
      </span>
      <input
        className="Input"
        style={
          Object {
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "resize": "both",
            "textAlign": undefined,
          }
        }
        type="text"
      />
    </div>
  </Connect>
</fieldset>
`;

exports[`TextField renders TextField with style 1`] = `
<fieldset
  className="TextField"
>
  <Connect>
    <input
      className="Input"
      style={
        Object {
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "resize": "both",
          "textAlign": "right",
        }
      }
      type="text"
    />
  </Connect>
</fieldset>
`;

exports[`TextField renders TextField with suffix 1`] = `
<fieldset
  className="TextField"
>
  <Connect>
    <div
      className="InputWrapper"
    >
      <input
        className="Input"
        style={
          Object {
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "resize": "both",
            "textAlign": undefined,
          }
        }
        type="text"
      />
      <span
        className="Suffix prefixclass"
      >
        <span>
          After Text
        </span>
      </span>
    </div>
  </Connect>
</fieldset>
`;
